br/pkg/restore/import.go

```diff
@@ -866,6 +851,276 @@ func (importer *FileImporter) downloadRawKVSST(
+
+	return downloadMetas, errDownload
+}
+
+func (importer *FileImporter) buildDownloadRequest(
+	file *backuppb.File,
+	rewriteRules *RewriteRules,
+	regionInfo *split.RegionInfo,
+	cipher *backuppb.CipherInfo,
+) (*import_sstpb.DownloadRequest, import_sstpb.SSTMeta, error) {
+	uid := uuid.New()
+	id := uid[:]
+	// Get the rewrite rule for the file.
+	fileRule := findMatchedRewriteRule(file, rewriteRules)
+	if fileRule == nil {
+		return nil, import_sstpb.SSTMeta{}, errors.Trace(berrors.ErrKVRewriteRuleNotFound)
+	}
+
+	// For the legacy version of TiKV, we need to encode the key prefix, since in the legacy
+	// version, the TiKV will rewrite the key with the encoded prefix without decoding the keys in
+	// the SST file. For the new version of TiKV that support keyspace rewrite, we don't need to
+	// encode the key prefix. The TiKV will decode the keys in the SST file and rewrite the keys
+	// with the plain prefix and encode the keys before writing to SST.
+
+	// for the keyspace rewrite mode
+	rule := *fileRule
+	// for the legacy rewrite mode
+	if importer.rewriteMode == RewriteModeLegacy {
+		rule.OldKeyPrefix = encodeKeyPrefix(fileRule.GetOldKeyPrefix())
+		rule.NewKeyPrefix = encodeKeyPrefix(fileRule.GetNewKeyPrefix())
+	}
+
+	sstMeta, err := GetSSTMetaFromFile(id, file, regionInfo.Region, &rule, importer.rewriteMode)
+	if err != nil {
+		return nil, import_sstpb.SSTMeta{}, err
+	}
+
+	req := &import_sstpb.DownloadRequest{
+		Sst:            *sstMeta,
+		StorageBackend: importer.backend,
+		Name:           file.GetName(),
+		RewriteRule:    rule,
+		CipherInfo:     cipher,
+		StorageCacheId: importer.cacheKey,
+		// For the older version of TiDB, the request type will  be default to `import_sstpb.RequestType_Legacy`
+		RequestType: import_sstpb.DownloadRequestType_Keyspace,
+		Context: &kvrpcpb.Context{
+			ResourceControlContext: &kvrpcpb.ResourceControlContext{
+				ResourceGroupName: "", // TODO,
+			},
+			RequestSource: kvutil.BuildRequestSource(true, kv.InternalTxnBR, kvutil.ExplicitTypeBR),
+		},
+	}
+	return req, *sstMeta, nil
+}
+
+func (importer *FileImporter) downloadSSTV2(
+	ctx context.Context,
+	regionInfo *split.RegionInfo,
+	files []*backuppb.File,
+	rewriteRules *RewriteRules,
+	cipher *backuppb.CipherInfo,
+	apiVersion kvrpcpb.APIVersion,
+) ([]*import_sstpb.SSTMeta, error) {
+	var mu sync.Mutex
+	downloadMetasMap := make(map[string]import_sstpb.SSTMeta)
+	resultMetasMap := make(map[string]*import_sstpb.SSTMeta)
+	downloadReqsMap := make(map[string]*import_sstpb.DownloadRequest)
+	for _, file := range files {
+		req, sstMeta, err := importer.buildDownloadRequest(file, rewriteRules, regionInfo, cipher)
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+		downloadMetasMap[file.Name] = sstMeta
+		downloadReqsMap[file.Name] = req
+	}
+
+	eg, ectx := errgroup.WithContext(ctx)
+	for _, p := range regionInfo.Region.GetPeers() {
+		peer := p
+		eg.Go(func() error {
+			importer.storeWorkerPoolRWLock.Lock()
```

Comment:
Use `RLock()` here to avoid lock blocken.
`If !ok, then` upgrade to use to `Lock()` because store change rarely occurs.

Reply:
```
importer.storeWorkerPoolRWLock.RLock()
workerCh, ok := importer.storeWorkerPoolMap[peer.GetStoreId()]
// handle the case that the store is new-scaled in the cluster
			if !ok {
				importer.storeWorkerPoolRWLock.UnRlock()
				importer.storeWorkerPoolRWLock.Lock()
                                 // Notice: worker channel can't replaced, because it is still used after unlock.
                                 if workerCh, ok = importer.storeWorkerPoolMap[peer.GetStoreId()]; !ok {
				        workerCh = utils.BuildWorkerTokenChannel(importer.concurrencyPerStore)
				    importer.storeWorkerPoolMap[peer.GetStoreId()] = workerCh
                                 }
				importer.storeWorkerPoolRWLock.UnLock()
			} else {
			        importer.storeWorkerPoolRWLock.UnRlock()
			}
```

Reply:
fixed